File: src/projects/hydroanalysis/compute_hydrovars.m
(* ::Package:: *)

(* compute_hydrovars.m *)
(* S Bernuzzi 10.2012 *)


(* VARIABLES *)

variables = {conD, pres, rho, epsl, v[a],
             g[a,b], alpha, beta[a], 
             xp, yp, zp,             
             Dh, Db, Du, ut,
             etot, uesc,
             vor[a], ivor[a], 
	     P[a], Pu[a]}


(* COMPUTE IN THIS ORDER *)

tocompute = {



    (* skip atm points *)

    Cinstruction == " if ((MATTER.USEMASK) && (mask[ijk]>.99)) {" ,

    Db == 0,
    Du == 0,  
    Dh == 0,
    ivor[a] == 0,
    
    Cinstruction == " continue;",
    Cinstruction == " }",
    


    (* 4-velocity component u_t *)
    
    vup[a] == v[a],    
    vsq == vup[a] vup[b] g[a,b],
    
    ooalpha == 1/alpha,
    Cinstruction == "if (alpha[ijk]<=ALPHAMIN){",
    Cinstruction == " printf(\" alpha<%e in hydroanalysis, reset\\n\",ALPHAMIN); ooalpha=1./ALPHAMIN;",    
    Cinstruction == "}",

    W == (1-vsq)^(-1/2),
    Cinstruction == "if (vsq>=1.) W = Wmax;", 

    uup[a] == W (vup[a] - beta[a] ooalpha),
    
    uupt == W ooalpha,

    betadown[b] == g[a,b] beta[a],

    udown[a] == g[a,b] uup[b] + betadown[a] uupt,
    
    h == 1+epsl+pres/rho,
    
    betasq ==  beta[b] betadown[b],
    betaduu == betadown[a] uup[a],

    udownt == (-alpha^2 + betasq ) uupt + betaduu,

    ut == udownt,

    ur     == xp vup[1] + yp vup[2] + zp vup[3],
    
    P[a] == conD h udown[a],    
    
    (* vorticity *)
    (* almost: it requires a derivative, done later *)

    ivor[a] == h udown[a],
    


    (* set various rest-masses *)
    
    Db   == conD,
    Du   == conD,
    Dh   == conD,

    (* compute bound/unbound rest-mass *)
    
    Cinstruction == " if ( (- udownt > 1)  && (ur > 0) ){" ,
    
    Db == 0,

    auupt == alpha uupt,    
    
    etot == conD*(h auupt - pres /(rho auupt)),
   
    uesc == conD epsl, 
   
    Pu[a] == P[a],    
 
    Cinstruction == " } else {",
    
    Du == 0,
    uesc == 0,
    etot == 0,
    Pu[a] == 0,

    Cinstruction == " } "

}


(**********************************************************************)
(* symmetries *)

g[a_,b_]       := g[b,a] /; !OrderedQ[{a,b}]
ginv[a_,b_]    := ginv[b,a] /; !OrderedQ[{a,b}]
delg[c_,a_,b_] := delg[c,b,a] /; !OrderedQ[{a,b}]
delgSST[c_,a_,b_] := delgSST[c,b,a] /; !OrderedQ[{a,b}]


(************************************************************************)
(* information for C output *)

(* information about the function 
   the C file will be functionname.c 
*)
includeANDdefine[] := Module[{},
  pr["#include \"bam.h\"\n"];
  pr["#include \"hydroanalysis.h\"\n"];
  pr["\n"];
  pr["#define Power(x,y) pow((double) (x), (double) (y))\n"];
  pr["#define Sqrt(x)    sqrt((double) (x))\n"];
  pr["#define Log(x)     log((double) (x))\n"];
  pr["#define pow2(x)    ((x)*(x))\n"];
  pr["#define pow4(x)    ((x)*(x)*(x)*(x))\n"];
  pr["#define pow2inv(x) (1.0/((x)*(x)))\n\n"];
  pr["#define Tanh(x)    tanh(x)\n"];
  pr["#define Sech(x)    (1/cosh(x))\n"];
  pr["#define Cal(x,y,z) ((x)?(y):(z))\n\n\n"];

  pr["#define ALPHAMIN 1e-6\n"]; 
  pr["\n\n\n"];
];

functionname = "compute_hydrovars";

functionarguments = "tVarList *u";

(* with OMP: *)
(* *)
functionbegin     = "bampi_openmp_start\n"; 
functionloop      = "forinnerpoints_ijk_openmp(level) {";
endfunctionloop   = "} endfor_ijk_openmp; /* loop i, j, k */";
endfunction       = "bampi_openmp_stop\n"; 
(* *)

(* without OMP: *)
(*
 functionloop      = "forinnerpoints_ijk(level) {";
 endfunctionloop   = "} endfor_ijk; /* loop i, j, k */";
*)



(* custom variable declaration
   we have to translate between the tensor names and the C variables  
   we may need special variables, for example for parameters
*)
variabledeclarations[] := Module[{},

  prdecvl[{conD, pres, rho, epsl, v[a],
           g[a,b], alpha, beta[a], 
           xp, yp, zp,             
           Dh, Db, Du,
           etot, uesc,
           vor[a], ivor[a],
	   P[a], Pu[a], ut}, 
          "u"];
                                        
  pr["tL *level = u->level;\n"];
  
  prdecvarname[{mask}, "matter_mask"];
  pr["\n"];
                    
  pr["const double Wmax = Getd(\"hydroanalysis_wMAX\");\n"];                    
                    
                    
];


(************************************************************************)
(* now we are ready to go *)


<< "../../math/MathToC/TensorEquationsToC.m"

File: src/projects/hydroanalysis/compute_hydrovars.c
/* compute_hydrovars.c */
/* Copyright (C) 1998 Bernd Bruegmann, 24.8.2023 */
/* Produced with Mathematica */

#include "bam.h"
#include "hydroanalysis.h"

#define Power(x,y) pow((double) (x), (double) (y))
#define Sqrt(x)    sqrt((double) (x))
#define Log(x)     log((double) (x))
#define pow2(x)    ((x)*(x))
#define pow4(x)    ((x)*(x)*(x)*(x))
#define pow2inv(x) (1.0/((x)*(x)))

#define Tanh(x)    tanh(x)
#define Sech(x)    (1/cosh(x))
#define Cal(x,y,z) ((x)?(y):(z))


#define ALPHAMIN 1e-6



void compute_hydrovars(tVarList *u)
{

double *conD = vldataptr(u, 0);
double *pres = vldataptr(u, 1);
double *rho = vldataptr(u, 2);
double *epsl = vldataptr(u, 3);
double *v1 = vldataptr(u, 4);
double *v2 = vldataptr(u, 5);
double *v3 = vldataptr(u, 6);
double *g11 = vldataptr(u, 7);
double *g12 = vldataptr(u, 8);
double *g13 = vldataptr(u, 9);
double *g22 = vldataptr(u, 10);
double *g23 = vldataptr(u, 11);
double *g33 = vldataptr(u, 12);
double *alpha = vldataptr(u, 13);
double *beta1 = vldataptr(u, 14);
double *beta2 = vldataptr(u, 15);
double *beta3 = vldataptr(u, 16);
double *xp = vldataptr(u, 17);
double *yp = vldataptr(u, 18);
double *zp = vldataptr(u, 19);
double *Dh = vldataptr(u, 20);
double *Db = vldataptr(u, 21);
double *Du = vldataptr(u, 22);
double *etot = vldataptr(u, 23);
double *uesc = vldataptr(u, 24);
double *vor1 = vldataptr(u, 25);
double *vor2 = vldataptr(u, 26);
double *vor3 = vldataptr(u, 27);
double *ivor1 = vldataptr(u, 28);
double *ivor2 = vldataptr(u, 29);
double *ivor3 = vldataptr(u, 30);
double *P1 = vldataptr(u, 31);
double *P2 = vldataptr(u, 32);
double *P3 = vldataptr(u, 33);
double *Pu1 = vldataptr(u, 34);
double *Pu2 = vldataptr(u, 35);
double *Pu3 = vldataptr(u, 36);
double *ut = vldataptr(u, 37);
tL *level = u->level;
int index_matter_mask = Ind("matter_mask");
double *mask = level->v[index_matter_mask + 0];

const double Wmax = Getd("hydroanalysis_wMAX");



bampi_openmp_start


double dx = level->dx;
double dy = level->dy;
double dz = level->dz;
double oo2dx = 1/(2*dx), oo2dy = 1/(2*dy), oo2dz = 1/(2*dz);
double oodx2 = 1/(dx*dx), oody2 = 1/(dy*dy), oodz2 = 1/(dz*dz);
double oo4dxdy = 1/(4*dx*dy);
double oo4dxdz = 1/(4*dx*dz);
double oo4dydz = 1/(4*dy*dz);

double auupt = 0.;
double betadown1 = 0.;
double betadown2 = 0.;
double betadown3 = 0.;
double betaduu = 0.;
double betasq = 0.;
double h = 0.;
double ooalpha = 0.;
double udown1 = 0.;
double udown2 = 0.;
double udown3 = 0.;
double udownt = 0.;
double ur = 0.;
double uup1 = 0.;
double uup2 = 0.;
double uup3 = 0.;
double uupt = 0.;
double vsq = 0.;
double vup1 = 0.;
double vup2 = 0.;
double vup3 = 0.;
double W = 0.;



forinnerpoints_ijk_openmp(level) {



 if ((MATTER.USEMASK) && (mask[ijk]>.99)) { 

Db[ijk]
=
0
;

Du[ijk]
=
0
;

Dh[ijk]
=
0
;

ivor1[ijk]
=
0
;

ivor2[ijk]
=
0
;

ivor3[ijk]
=
0
;


 continue; 


 } 

vup1
=
v1[ijk]
;

vup2
=
v2[ijk]
;

vup3
=
v3[ijk]
;

vsq
=
2.*(vup1*(vup2*g12[ijk] + vup3*g13[ijk]) + vup2*vup3*g23[ijk]) + 
  g11[ijk]*pow2(vup1) + g22[ijk]*pow2(vup2) + g33[ijk]*pow2(vup3)
;

ooalpha
=
1/alpha[ijk]
;


if (alpha[ijk]<=ALPHAMIN){ 


 printf(" alpha<%e in hydroanalysis, reset\n",ALPHAMIN); ooalpha=1./ALPHAMIN; 


} 

W
=
1/Sqrt(1. - vsq)
;


if (vsq>=1.) W = Wmax; 

uup1
=
W*(vup1 - ooalpha*beta1[ijk])
;

uup2
=
W*(vup2 - ooalpha*beta2[ijk])
;

uup3
=
W*(vup3 - ooalpha*beta3[ijk])
;

uupt
=
ooalpha*W
;

betadown1
=
beta1[ijk]*g11[ijk] + beta2[ijk]*g12[ijk] + beta3[ijk]*g13[ijk]
;

betadown2
=
beta1[ijk]*g12[ijk] + beta2[ijk]*g22[ijk] + beta3[ijk]*g23[ijk]
;

betadown3
=
beta1[ijk]*g13[ijk] + beta2[ijk]*g23[ijk] + beta3[ijk]*g33[ijk]
;

udown1
=
betadown1*uupt + uup1*g11[ijk] + uup2*g12[ijk] + uup3*g13[ijk]
;

udown2
=
betadown2*uupt + uup1*g12[ijk] + uup2*g22[ijk] + uup3*g23[ijk]
;

udown3
=
betadown3*uupt + uup1*g13[ijk] + uup2*g23[ijk] + uup3*g33[ijk]
;

h
=
1. + epsl[ijk] + pres[ijk]/rho[ijk]
;

betasq
=
betadown1*beta1[ijk] + betadown2*beta2[ijk] + betadown3*beta3[ijk]
;

betaduu
=
betadown1*uup1 + betadown2*uup2 + betadown3*uup3
;

udownt
=
betaduu + uupt*(betasq - pow2(alpha[ijk]))
;

ut[ijk]
=
udownt
;

ur
=
vup1*xp[ijk] + vup2*yp[ijk] + vup3*zp[ijk]
;

P1[ijk]
=
h*udown1*conD[ijk]
;

P2[ijk]
=
h*udown2*conD[ijk]
;

P3[ijk]
=
h*udown3*conD[ijk]
;

ivor1[ijk]
=
h*udown1
;

ivor2[ijk]
=
h*udown2
;

ivor3[ijk]
=
h*udown3
;

Db[ijk]
=
conD[ijk]
;

Du[ijk]
=
conD[ijk]
;

Dh[ijk]
=
conD[ijk]
;


 if ( (- udownt > 1)  && (ur > 0) ){ 

Db[ijk]
=
0
;

auupt
=
uupt*alpha[ijk]
;

etot[ijk]
=
conD[ijk]*(auupt*h - pres[ijk]/(auupt*rho[ijk]))
;

uesc[ijk]
=
conD[ijk]*epsl[ijk]
;

Pu1[ijk]
=
P1[ijk]
;

Pu2[ijk]
=
P2[ijk]
;

Pu3[ijk]
=
P3[ijk]
;


 } else { 

Du[ijk]
=
0
;

uesc[ijk]
=
0
;

etot[ijk]
=
0
;

Pu1[ijk]
=
0
;

Pu2[ijk]
=
0
;

Pu3[ijk]
=
0
;


 }  



} endfor_ijk_openmp; /* loop i, j, k */



bampi_openmp_stop


}  /* function */

/* compute_hydrovars.c */
/* nvars = 38, nauxs = 22, n* = 129,  n/ = 26,  n+ = 43, n = 198, O = 0 */
